src/react-cancelable.js

```diff
@@ -0,0 +1,80 @@
+import React from "react";
+import ReactDOM from "react-dom";
+import { Observable } from "rx";
+
+const inTestingEnv = typeof jasmine !== "undefined";
+
+export function Cancelable() {
+	return function decorator(DecoratedComponent) {
+		return class Disposable extends React.Component {
+			disposables = [];
+			render() {
+				return (
+					<DecoratedComponent
+						{...this.props}
+						cancelWhenUnmounted={this.cancelWhenUnmounted}
+						cancelAllSubscriptions={this.cancelAllSubscriptions}
+					/>
+				);
+			}
+			componentWillUnmount() {
+				this.cancelAllSubscriptions();
+			}
+			cancelWhenUnmounted = cancelWhenUnmounted.bind(this);
+			cancelAllSubscriptions = cancelAllSubscriptions.bind(this);
+		};
+	};
+}
+
+export function cancelableModal(El, props = {}) {
+	let el;
+
+	return Observable.create(observer => {
+		el = document.createElement("div");
+		document.body.appendChild(el);
+
+		ReactDOM.render(
+			<El {...props} close={observer.onCompleted} save={observer.onNext} />,
+			el
+		);
+	}).finally(() => {
+		ReactDOM.unmountComponentAtNode(el);
+		el.parentNode.removeChild(el);
+	});
+}
+
+export class CancelableComponent extends React.Component {
+	disposables = [];
+	render() {
+		return this.props.children(
+			this.cancelWhenUnmounted,
+			this.cancelAllSubscriptions
+		);
+	}
+	componentWillUnmount() {
+		this.cancelAllSubscriptions();
+	}
+	cancelWhenUnmounted = cancelWhenUnmounted.bind(this);
+	cancelAllSubscriptions = cancelAllSubscriptions.bind(this);
+}
+
+function cancelWhenUnmounted(...thingsToCancel) {
+	thingsToCancel.forEach(thingToCancel => {
+		if (!thingToCancel || typeof thingToCancel.dispose !== "function") {
```

Comment:
it might be nice to try to detect if the user has provided an observable instead of a subscription/disposable. And give them a helpful error message in that case that is different than the normal one. Something like `cancelWhenUnmounted should be called with a subscription, not an observable. Did you forget to subscribe to your observable?`

It's something I've seen a few times in our code at Canopy

Reply:
Good idea